quickwit/quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs

```diff
@@ -543,18 +781,57 @@ fn build_mapping_tree_from_entries<'a>(
     entries: &'a [FieldMappingEntry],
     field_path: &mut Vec<&'a str>,
     schema: &mut SchemaBuilder,
-) -> anyhow::Result<MappingNode> {
+) -> anyhow::Result<(MappingNode, Vec<Field>)> {
     let mut mapping_node = MappingNode::default();
+    let mut concatenate_fields = Vec::new();
+    let mut concatenate_dynamic_fields = Vec::new();
     for entry in entries {
-        field_path.push(&entry.name);
-        if mapping_node.branches.contains_key(&entry.name) {
-            bail!("duplicated field definition `{}`", entry.name);
+        if let FieldMappingType::Concatenate(_) = &entry.mapping_type {
+            concatenate_fields.push(entry);
+        } else {
+            field_path.push(&entry.name);
+            if mapping_node.branches.contains_key(&entry.name) {
+                bail!("duplicated field definition `{}`", entry.name);
+            }
+            let (child_tree, mut dynamic_fields) =
+                build_mapping_from_field_type(&entry.mapping_type, field_path, schema)?;
+            field_path.pop();
+            mapping_node.insert(&entry.name, child_tree);
+            concatenate_dynamic_fields.append(&mut dynamic_fields);
         }
-        let child_tree = build_mapping_from_field_type(&entry.mapping_type, field_path, schema)?;
-        field_path.pop();
-        mapping_node.insert(&entry.name, child_tree);
     }
-    Ok(mapping_node)
+    for concatenate_field_entry in concatenate_fields {
+        let FieldMappingType::Concatenate(options) = &concatenate_field_entry.mapping_type else {
+            continue; // unreachable
+        };
+        let name = &concatenate_field_entry.name;
+        if mapping_node.branches.contains_key(name) {
+            bail!("duplicated field definition `{}`", name);
+        }
+        let text_options: TextOptions = options.clone().into();
+        let field = schema.add_text_field(name, text_options);
+        for sub_field in &options.concatenate_fields {
+            for matched_field in
+                mapping_node
+                    .find_field_mapping_leaf(sub_field)
```

Comment:
What happens for json field?

If I have a json field called attributes and I concat
`attributes.color`, is this an error, or is not an error but working in strange ways?

Reply:
If the spec is acceptable, can we have a unit test for this?

Reply:
right now, it's an error at index creation time. In the future, I hope it may become possible to implement. I've added a commented-out test